### PR TITLE
addressing: add support for Address Map resource

### PR DESCRIPTION
**Pull request by hunts**
## Description

This pull request adds support for the public `Address Map` apis.

About `Address Map`: https://developers.cloudflare.com/byoip/proxy-ips/
The API docs: https://api.cloudflare.com/#ip-address-management-address-maps-properties

## Types of changes

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] This change is using publicly documented (api.cloudflare.com or developers.cloudflare.com) and stable APIs.

### CODE REVIEW

File: addressing_address_map.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	"fmt"
+	"net/http"
+	"time"
+)
+
+// AddressMap contains information about an address map.
+type AddressMap struct {
+	ID           string                 `json:"id"`
+	Description  string                 `json:"description"`
+	DefaultSNI   string                 `json:"default_sni"`
```
Comment (aeriff) [inline addressing_address_map.go]:
```suggestion
	DefaultSNI   *string                 `json:"default_sni"`
```

Same here, this field can be null.

File: addressing_address_map.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+// AddressMap contains information about an address map.
+type AddressMap struct {
+	ID           string                 `json:"id"`
+	Description  string                 `json:"description"`
```
Comment (aeriff) [inline addressing_address_map.go]:
```suggestion
	Description  *string                 `json:"description"`
```
Sorry that the docs are not explicit about this (we need to fix it), but this field can be null.

File: addressing_address_map.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	DefaultSNI   string                 `json:"default_sni"`
+	Enabled      bool                   `json:"enabled"`
+	CanDelete    bool                   `json:"can_delete"`
+	CanModifyIPs bool                   `json:"can_modify_ips"`
+	Memberships  []AddressMapMembership `json:"memberships"`
+	IPs          []AddressMapIP         `json:"ips"`
+	CreatedAt    *time.Time             `json:"created_at"`
```
Comment (aeriff) [inline addressing_address_map.go]: The timestamp fields on all Address Maps endpoints are always non-null, but I am not sure if this is convention in cloudflare-go so I leave it up to you.

Comment (hunts) [reply addressing_address_map.go]: There are mixed usages of pointers and values for time fields in this repo. I'll make the change if the api does not return null.

Comment (jacobbednarz) [reply addressing_address_map.go]: `time.Time` usage is definitely varied (sorry!). i'm working on documenting type conventions better as people are commonly tripping up on uninitialised values and zero values with this library and i think i can do a better job of explaining where the traps are.

we have [builtin in helpers for type conversion](https://github.com/cloudflare/cloudflare-go/blob/master/convert_types.go) so this should be easy for most to adopt.

Comment (jacobbednarz) [reply addressing_address_map.go]: similarly, i recommend all booleans are boolean pointers so avoid some issues with Terraform.

File: addressing_address_map.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+type AddressMapPatch struct {
+	Description *string `json:"description,omitempty"`
+	Enabled     *bool   `json:"enabled,omitempty"`
+	DefaultSNI  *string `json:"default_sni,omitempty"`
+}
+
+// ListAddressMaps lists all address maps owned by the account.
+//
+// API reference: https://developers.cloudflare.com/api/operations/ip-address-management-address-maps-list-address-maps
+func (api *API) ListAddressMaps(ctx context.Context, accountID string) ([]AddressMap, error) {
```
Comment (jacobbednarz) [inline addressing_address_map.go]: i've outlined some conventions for new functionality at https://github.com/cloudflare/cloudflare-go/blob/master/docs/experimental.md. in essence:

- all methods have max 3 parameters
- the third parameter is always a struct (or a single ID for `Get` methods)
- the second parameter is a `ResourceContainer`
- where possible, we stick to `Get`, `List`, `Update`, `Create` and `Delete` method prefixes for consistency
- `List` operations automatically paginate all results

you can check out https://github.com/cloudflare/cloudflare-go/blob/master/tunnel.go for prior art if you're wondering how it all hangs together.

Comment (hunts) [reply addressing_address_map.go]: I've made some changes according to these experimental items. The list api itself in this case does not support paging. Do I need to do anything with it?

File: addressing_address_map.go (hunk 3 above)
Comment (hunts) [reply addressing_address_map.go]: Are you suggesting make all time and boolean fields pointers? And can you please elaborate what would be the issues Terraform run into with non-null fields? 

> `i recommend all booleans are boolean pointers so avoid some issues with Terraform`

**Comment by hunts** [comment]
The `Coverage` action seems to be broken. @jacobbednarz . Is it a known issue?

**Comment by jacobbednarz** [comment]
I'll have to have a look into it as to where the comparison is happening but that's fine to ignore it for now.

File: addressing_address_map.go (hunk 3 above)
Comment (jacobbednarz) [reply addressing_address_map.go]: that would be my suggestion, yes. on the terraform side, there needs to be a way to determine if a field is uninitialised, false or true and the only reliable way of doing that is with pointer fields.

File: addressing_address_map.go (hunk 4 above)
Comment (jacobbednarz) [reply addressing_address_map.go]: as long as the signature matches, that's fine. we have plenty of APIs that don't paginate but need to look the same.

File: addressing_address_map.go (hunk 3 above)
Comment (jacobbednarz) [reply addressing_address_map.go]: i've gone ahead and pushed up some changes to address this

**Comment by jacobbednarz** [comment]
thanks for this one @hunts 🍰

**Comment by hunts** [comment]
Awesome! Thanks for the help @jacobbednarz on releasing this feature.

**Pull request merged**